http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcRequestAuthenticator.java

```diff
@@ -416,4 +482,111 @@ private static boolean hasScope(String scopeParam, String targetScope) {
+        forwardedQueryParams.add(new BasicNameValuePair(RESPONSE_TYPE, CODE));
+        forwardedQueryParams.add(new BasicNameValuePair(CLIENT_ID, deployment.getResourceName()));
+
+        JwtClaims jwtClaims = new JwtClaims();
+        jwtClaims.setIssuer(deployment.getIssuerUrl());
+        jwtClaims.setAudience(deployment.getProviderUrl());
+        for ( NameValuePair parameter: forwardedQueryParams) {
+            jwtClaims.setClaim(parameter.getName(), parameter.getValue());
+        }
+
+        // sign JWT first before encrypting
+        JsonWebSignature signedRequest = signRequest(jwtClaims);
+
+        // Encrypting optional
+        if (deployment.getRequestEncryptAlgorithm() != null && !deployment.getRequestEncryptAlgorithm().isEmpty() &&
+            deployment.getRequestEncryptEncValue() != null && !deployment.getRequestEncryptEncValue().isEmpty()) {
+            return encryptRequest(signedRequest).getCompactSerialization();
+        } else {
+            return signedRequest.getCompactSerialization();
+        }
+    }
+
+    public KeyPair getkeyPair() throws IOException {
+        if (deployment.getClientKeyStoreFile().contains(PROTOCOL_CLASSPATH)) {
+            deployment.setClientKeyStoreFile(deployment.getClientKeyStoreFile().replace(PROTOCOL_CLASSPATH, Objects.requireNonNull(Thread.currentThread().getContextClassLoader().getResource("")).getPath()));
+        }
+        if (!deployment.getRequestSignatureAlgorithm().equals(NONE) && deployment.getClientKeyStoreFile() == null){
+            throw log.invalidKeyStoreConfiguration();
+        } else {
+            return loadKeyPairFromKeyStore(deployment.getClientKeyStoreFile(),
+                    deployment.getClientKeyStorePassword(), deployment.getClientKeyPassword(),
+                    deployment.getClientKeyAlias(), deployment.getClientKeystoreType());
+        }
+    }
+
+    public JsonWebSignature signRequest(JwtClaims jwtClaims) throws Exception {
+        JsonWebSignature jsonWebSignature = new JsonWebSignature();
+        jsonWebSignature.setPayload(jwtClaims.toJson());
+
+        if (deployment.getRequestSignatureAlgorithm().contains(NONE)) { //unsigned
+            jsonWebSignature.setAlgorithmConstraints(AlgorithmConstraints.NO_CONSTRAINTS);
+            jsonWebSignature.setAlgorithmHeaderValue(NONE);
+        } else if (deployment.getRequestSignatureAlgorithm().contains(HS_256)
+                || deployment.getRequestSignatureAlgorithm().contains(HS_384)
+                || deployment.getRequestSignatureAlgorithm().contains(HS_512)) { //signed with symmetric key
+            jsonWebSignature.setAlgorithmHeaderValue(deployment.getRequestSignatureAlgorithm());
+            Key key = new HmacKey(deployment.getResourceCredentials().get("secret").toString().getBytes(StandardCharsets.UTF_8));   //the client secret is a shared secret between the server and the client
+            jsonWebSignature.setDoKeyValidation(false);
```

Comment:
@PrarthonaPaul Can they use JWT as a client credential? Can you please describe the background for the situation you are asking about? 

in the proposal you mentioned: 

> In order to sign the jwt using an algorithm other than none, the user must specify the KeyPair used. 

So I thought the keypair can be used in all situations for signing but you might be asking about something else

Reply:
Yes, sorry. 
According to this doc: https://www.keycloak.org/docs/latest/securing_apps/#_client_authentication_adapter
we can either set up the client credentials to be just the client secret, which is a just a hash/SHA (like 19666a4f-32dd-4049-b082-684c74115f28). 
But you can also configure it using a JWT. But it is not related to the request RFE. 
This is what the options are for specifying the client credentials: 
![image](https://github.com/wildfly-security/wildfly-elytron/assets/83880425/d0fbc0dc-2a35-4e66-a4ed-e0e2de1e94b0)

Now the reason I am using the client secret to sign a JWT for my RFE is because as we found out HS keys are symmetric. And from what I have seen instead of creating a certificate, we create a secret that is shared between the client. For Keycloak, the client secret is something that is shared between the client and the server. 

But what I am wondering is what happens when the user configures the client credentials as a JWT or X509 certificate?